File: app/src/main/java/org/ar25/androidno/ui/OnLoadMoreEndlessRecyclerView.java
package org.ar25.androidno.ui;


import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;


//
// Thanks to: https://gist.github.com/ssinss/e06f12ef66c51252563e
//
public abstract class OnLoadMoreEndlessRecyclerView extends RecyclerView.OnScrollListener {
    private int previousTotal = 0; // The total number of items in the dataset after the last load
    private boolean loading = true; // True if we are still waiting for the last set of data to load.
    private int visibleThreshold = 1; // The minimum amount of items to have below your current scroll position before loading more.

    private LinearLayoutManager mLinearLayoutManager;

    public OnLoadMoreEndlessRecyclerView(LinearLayoutManager linearLayoutManager) {
        mLinearLayoutManager = linearLayoutManager;
    }

    @Override
    public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
        super.onScrolled(recyclerView, dx, dy);

        int visibleItemCount = recyclerView.getChildCount();
        int totalItemCount = mLinearLayoutManager.getItemCount();
        int firstVisibleItem = mLinearLayoutManager.findFirstVisibleItemPosition();

        if (loading) {
            if (totalItemCount > previousTotal) {
                loading = false;
                previousTotal = totalItemCount;
            }
        }
        if (!loading && (totalItemCount - visibleItemCount)
                <= (firstVisibleItem + visibleThreshold)) {
            // End has been reached

            onLoadMore();

            loading = true;
        }
    }

    public abstract void onLoadMore();
}

File: app/src/main/java/org/ar25/androidno/ui/OnLoadMoreEndlessRecyclerView.kt
package org.ar25.androidno.ui


import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView


//
// Thanks to: https://gist.github.com/ssinss/e06f12ef66c51252563e
//
open class OnLoadMoreEndlessRecyclerView(
        private val linearLayoutManager: LinearLayoutManager,
        private val onLoadMore: () -> Unit

    ) : RecyclerView.OnScrollListener() {

    private var previousTotal = 0 // The total number of items in the dataset after the last load
    private var loading = true // True if we are still waiting for the last set of data to load.
    private val visibleThreshold = 1 // The minimum amount of items to have below your current scroll position before loading more.


    override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
        super.onScrolled(recyclerView, dx, dy)

        val visibleItemCount = recyclerView.childCount
        val totalItemCount = linearLayoutManager.itemCount
        val firstVisibleItem = linearLayoutManager.findFirstVisibleItemPosition()

        if (loading) {
            if (totalItemCount > previousTotal) {
                loading = false
                previousTotal = totalItemCount
            }
        }
        if (!loading && totalItemCount - visibleItemCount <= firstVisibleItem + visibleThreshold) {
            // End has been reached

            onLoadMore()

            loading = true
        }
    }
}
